benchmarks/pytorch-lightning/main.py

```diff
@@ -0,0 +1,172 @@
+from pytorch_lightning import Callback, LightningModule, Trainer
+from simple_parsing import ArgumentParser
+from typing_extensions import ParamSpec
+from utils import DataOptions
+
+P = ParamSpec("P")
+
+
+def main(
+    trainer_type: type[Callable[P, Trainer]] = Trainer,
+    *trainer_default_args: P.args,
+    **trainer_default_kwargs: P.kwargs,
+):
+    """
+    Runs a PyTorch-Lightning benchmark.
+    
+    The benchmark consists of training a model of type `model_type`, using a `Trainer`, on the 
+    `LightningDataModule` that is created from a `DataOptions` that is also parsed from the
+    command-line.
+
+    To create the `Trainer`, the values from `trainer_default_args` and `trainer_default_kwargs`
+    are used as the defaults. The values parsed from the command-line then overwrite these values.
+
+    NOTE: trainer_type is just used so we can get nice type-checks for the trainer kwargs. We
+    don't actually expect a different subclass of Trainer to be passed.
+    """
+    # Create a parser so we can extract all the args of the Trainer class, and add it to our own,
+    # (better) ArgumentParser.
+    trainer_parser = ArgumentParser(add_help=False)
+    trainer_parser = Trainer.add_argparse_args(trainer_parser)
+
+    # Copy the arguments from the Trainer parser, and then add our own.
+    parser = ArgumentParser(parents=[trainer_parser], add_help=True)
+
+    trainer_parser.set_defaults(enable_checkpointing=False)
+
+    trainer_signature = inspect.signature(Trainer)
+    if trainer_default_kwargs:
+        # Overwrite the default values with those from the `trainer_defaults` dict.
+        sig = trainer_signature.bind_partial(**trainer_default_kwargs)
+        print(f"Overwriting default values for the Trainer: {sig.arguments}")
+        trainer_parser.set_defaults(**sig.arguments)
+
+    # Add the arguments for the Model:
+    parser.add_arguments(Model.HParams, "hparams")
+
+    # Add arguments for the dataset choice / setup:
+    parser.add_arguments(DataOptions, "options")
+
+    args = parser.parse_args()
+    args_dict = vars(args)
+
+    # NOTE: Instead of creating the Trainer from the args directly, we instead create it from the
+    # kwargs, so we can have better control over some args like `callbacks`, `logger`, etc that
+    # accept objects.
+    # trainer = Trainer.from_argparse_args(args)
+
+    hparams: Model.HParams = args_dict.pop("hparams")
+    options: DataOptions = args_dict.pop("options")
+
+    print(f"HParams: \n{hparams.dumps_yaml()}")
+    print(f"Options: \n{options.dumps_yaml()}")
+
+    # Rest of `args_dict` is only for the Trainer.
+    trainer_kwargs = trainer_default_kwargs.copy()
+    trainer_kwargs.update(**args_dict)
+    callbacks = trainer_kwargs.setdefault("callbacks", [])
+    assert isinstance(callbacks, list)
+    if not any(isinstance(c, GivingCallback) for c in callbacks):
+        callbacks.append(GivingCallback())
+
+    print(f"Trainer kwargs:")
+    import pprint
```

Comment:
Hoist imports to top, unless exceptional circumstances.

Reply:
Fixed in 7f3383a